### PR TITLE
Add expert dropout

### PR DESCRIPTION
**Pull request by samhavens**
From the Switch paper, "We thus propose a simple way to alleviate [MoE overfitting] during fine tuning: increase the dropout inside the experts, which we name as expert dropout. During fine-tuning we simply increase the dropout rate by a significant amount only at the interim feed-forward computation at each expert layer."

### CODE REVIEW

File: megablocks/layers/moe.py [hunk 1]
```diff
@@ -200,6 +204,11 @@ def permute_and_compute(
             x, indices, expert_weights, bins, top_k)
 
     def forward_once(self, x, expert_weights, top_experts):
+        # Expert dropout: Randomly drop experts based on the dropout rate
```
Comment (tgale96) [inline megablocks/layers/moe.py:207]: I think that you could add this in the `forward` function and avoid duplicating it for both `forward_once` and `parallel_forward_once`?

File: megablocks/layers/moe.py [hunk 2]
```diff
@@ -200,6 +204,11 @@ def permute_and_compute(
...
+        if self.training and self.args.expert_dropout_rate > 0:
+            dropout_mask = torch.rand_like(expert_weights) < self.args.moe_expert_dropout_rate
+            top_experts = top_experts.masked_fill(dropout_mask, -1)
```
Comment (tgale96) [inline megablocks/layers/moe.py:210]: I'm not certain that assigned these tokens to expert '-1' will have the desired effect here...

Also, I'd understood that the "expert dropout" method from Switch Transformer was just turning up the dropout rate in the FFN, rather than dropping whole experts? Is this something different?

Comment (samhavens) [reply megablocks/layers/moe.py:210]: I just looked at the paper and you are right. I overfit to the name, sorry!
